File: java/leetcode-pratice/leetcode-pratice/链表/两数相加2.java
package com.boshrong.leetcode.链表;

public class 两数相加2 {


    public static void main(String[] args) {

    }
    public  static ListNode addTwoNumbers(ListNode l1, ListNode l2) {
        ListNode headNode = new ListNode(0);
        int n1,n2,carry=0;
        ListNode current=headNode;
        while(l1!=null||l2!=null||carry!=0){
            if(l1==null){
                n1=0;
            }else{
                n1=l1.val;
                l1=l1.next;
            }
            if(l2==null){
                n2=0;
            }else{
                n2= l2.val;
                l2=l2.next;
            }
            //要创建一个新结点
            current.next=new ListNode((n1+n2+carry)%10);
            current=current.next;
            carry=(n1+n2+carry)/10;
        }
        return  headNode.next;


    }
}

File: go/LinkedList/两数相加2.go
package LinkedList

/**
 * Definition for singly-linked list.
 */
type ListNode struct {
	Val  int
	Next *ListNode
}

func addTwoNumbers(l1 *ListNode, l2 *ListNode) *ListNode {
	head := &ListNode{Val: 0}
	num1, num2 := 0, 0
	carry := 0
	// define work Point
	current := head
	for l1 != nil || l2 != nil || carry != 0 {
		// l1 or l2 maybe nil
		if l1 == nil {
			num1 = 0
		} else {
			num1 = l1.Val
			l1 = l1.Next
		}
		if l2 == nil {
			num2 = 0
		} else {
			num2 = l2.Val
			l2 = l2.Next
		}
		l := &ListNode{(num1 + num2 + carry) % 10, nil}
		carry = (num1 + num2 + carry) / 10

		// post insert
		current.Next = l
		current = l
	}
	return head.Next
}
